lib/graph_gateway.go

```diff
@@ -0,0 +1,482 @@
+package lib
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/filecoin-saturn/caboose"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	leveldb "github.com/ipfs/go-ds-leveldb"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	exchange "github.com/ipfs/go-ipfs-exchange-interface"
+	"github.com/ipfs/go-libipfs/blocks"
+	"github.com/ipfs/go-libipfs/files"
+	"github.com/ipfs/go-libipfs/gateway"
+	"github.com/ipfs/go-namesys"
+	"github.com/ipfs/go-namesys/resolve"
+	ipfspath "github.com/ipfs/go-path"
+	nsopts "github.com/ipfs/interface-go-ipfs-core/options/namesys"
+	ifacepath "github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/ipld/go-car"
+	routinghelpers "github.com/libp2p/go-libp2p-routing-helpers"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"go.uber.org/multierr"
+	"io"
+	"net/http"
+)
+
+// type DataCallback = func(resource string, reader io.Reader) error
+// TODO: Don't use a caboose type, perhaps ask them to use a type alias instead of a type
+type DataCallback = caboose.DataCallback
+
+type CarFetcher interface {
+	Fetch(ctx context.Context, path string, cb DataCallback) error
```

Comment:
:thought_balloon:  `path` mixes content path with CAR format params. 

What if we replace `path` with `ImmutablePath` and have a separate `params url.Values` with  optional `depth` and `bytes` ?

Reply:
yeah, this is pretty messy. Also I got it wrong here (not that it matters because there's no behavior change) since Caboose's implementation of `Fetch` uses the CAR accept header anyway.

As described [above](https://github.com/ipfs/bifrost-gateway/pull/61#discussion_r1147564229) we should clean up and define the abstractions we want here and then pass them onto Caboose or our HTTP proxy calls.

We can use `url.Values` or be more specific with types, likely doesn't matter much 🤷.

Reply:
I think we need something specific here because `url.Values` is `map` and golang won't preserve order. 
This means requests for the same thing may have params in different order, and caches that use URL as cache key will miss due to different URLs.